packages/compiler/src/server/tspconfig/completion.ts

```diff
@@ -204,10 +319,153 @@ export async function provideTspconfigCompletionItems(
+                    documentation: cur.description,
+                  });
+                }
+              }
+            } else {
+              // built-in variables
+              result.push(
+                ...["cwd", "project-root"].map((value) => {
+                  const item: CompletionItem = {
+                    label: value,
+                    kind: CompletionItemKind.Value,
+                    documentation: cur.description,
+                  };
+                  return item;
+                }),
+              );
+              // parameters
+              for (const param of target.parameters) {
+                if (!nodePath.includes(param)) {
+                  result.push({
+                    label: param,
+                    kind: CompletionItemKind.Value,
+                    documentation: cur.description,
+                  });
+                }
+              }
+            }
+          } else if (cur.type === "string") {
+            // extends
+            if (relativeFiles.length > 0) {
+              for (const file of relativeFiles) {
+                if (target.siblings.includes(file)) {
+                  continue;
+                }
+
+                const item: CompletionItem = {
+                  label: file,
+                  kind: CompletionItemKind.Field,
+                };
+                result.push(item);
+              }
+            }
           }
         }
       });
 
     return distinctArray(result, (t) => t.label);
   }
 }
+
+/**
+ * Get the new text value
+ * @param sourceType input quote type(single, double, none)
+ * @param formatText format text value
+ * @returns
+ */
+function getNewTextValue(sourceType: string, formatText: string): string {
+  let newText: string = "";
+  if (sourceType === "QUOTE_SINGLE") {
+    newText = `${formatText}'`;
+  } else if (sourceType === "QUOTE_DOUBLE") {
+    newText = `${formatText}"`;
+  } else {
+    newText = `"${formatText}"`;
+  }
+  return newText;
+}
+
+/**
+ * Get the position of the new text
+ * @param newText  the new text
+ * @param source source text
+ * @param tspConfigPosition original position
+ * @returns
+ */
+function getNewTextAndPosition(
+  newText: string,
+  source: string,
+  tspConfigPosition: Position,
+): TextEdit {
+  return TextEdit.replace(
+    Range.create(
+      Position.create(tspConfigPosition.line, tspConfigPosition.character - source.length),
+      Position.create(tspConfigPosition.line, tspConfigPosition.character + newText.length),
+    ),
+    newText,
+  );
+}
+
+/**
+ * Find a set of relative paths to files with the same suffix
+ * @param rootPath The root path of the current configuration file
+ * @param fileExtension  File extension
+ * @param log log function
+ * @param configFile Configuration file name
+ * @returns
+ */
+function findFilesWithSameExtension(
+  rootPath: string,
+  fileExtension: string,
+  log: (log: ServerLog) => void,
+  configFile: string = "",
+): string[] {
```

Comment:
the method feels complex to me. isn't it something like below psudo-code?
(await readdir(pth, {withFileTypes: true}))
    .filter(d => d.isDirectory() || d is file and endsWith extension)
    .map(d => get file/dirname);